### PR TITLE
Endpoint healthz reported via command & in endpoint get

### PR DESCRIPTION
**Pull request by raybejjani**
**Summary of changes**:
Added healthz to `endpoint get` and introduced `endpoint healthz`. It pretty prints and returns json

@tgraf I guessed on some of this logic, let me know if it is incorrect.

Sample output:
```
$ cilium endpoint healthz 175 --json
{
  "bpf": "OK",
  "connected": true,
  "health": "OK",
  "policy": "OK"
}

$ cilium endpoint healthz 175
Health:      OK
BPF:         OK
Policy:      OK
Connected:   yes
```

Fixes: #1277 

```release-note
Added cilium endpoint healthz command that returns a summary of the endpoint's health
```

**How to test (optional)**:
Tests pass.
make clean && make generate-api && make reload
Note: There might be new/old files left in the api directory. In that case:
rm -rf ./api/ && git checkout -f

### CODE REVIEW

**Comment by aanm** [comment]
When an endpoint will be showed as `Connected:   no`?

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
...
+
+	case models.EndpointStateNotReady:
+		bpfStatus = models.EndpointHealthzStatusWarning
+
+	case models.EndpointStateDisconnected:
+		bpfStatus = models.EndpointHealthzStatusDisabled
+	}
+
+	policyStatus := models.EndpointHealthzStatusOK
+	switch currentState {
```
Comment (aanm) [inline pkg/endpoint/endpoint.go]: Why do you have a duplicate switch? You can assign `policyStatus` and `bpfStatus` at the same time in a single switch

File: pkg/endpoint/endpoint.go [hunk 2]
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
...
+
+	case models.EndpointStateCreating:
+		// Does this rely on default deny or default allow?
+		policyStatus = models.EndpointHealthzStatusDisabled
+
+	case models.EndpointStateWaitingForIdentity:
+		policyStatus = models.EndpointHealthzStatusBootstrap
+
+	case models.EndpointStateNotReady:
+		bpfStatus = models.EndpointHealthzStatusWarning
```
Comment (aanm) [inline pkg/endpoint/endpoint.go]: `s/bpfStatus/policyStatus`

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
...
+
+	case models.EndpointStateDisconnected:
+		bpfStatus = models.EndpointHealthzStatusDisabled
```
Comment (aanm) [inline pkg/endpoint/endpoint.go]: `s/bpfStatus/policyStatus`

File: pkg/client/endpoint.go [hunk 4]
```diff
@@ -74,6 +74,16 @@ func (c *Client) EndpointLogGet(id string) (models.EndpointStatusLog, error) {
 	return resp.Payload, nil
 }
 
+// EndpointLogHealthz returns endpoint healthz
```
[the remark this answers is not in the document]
Comment (aanm) [reply pkg/client/endpoint.go]: `s/EndpointLogHealthz/EndpointHealthzGet`

File: pkg/endpoint/endpoint.go [hunk 5]
```diff
@@ -453,7 +453,7 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
 // newHealthz returns the endpoint's healthz object.
 //
 // Must be called with e.Mutex locked.
-func (e *Endpoint) getHealthzModel() *models.EndpointHealthz {
+func (e *Endpoint) GetHealthzModel() *models.EndpointHealthz {
```
Comment (aanm) [inline pkg/endpoint/endpoint.go]: Create a public method that can be called without the e.Mutex locked. You are calling this function without locking the endpoint in daemon/endpoint.go:581

**Review by aanm — CHANGES_REQUESTED** [review]

File: api/v1/openapi.yaml [hunk 6]
```diff
@@ -755,6 +755,9 @@ definitions:
       status:
         description: Most recent status log. See endpoint/{id}/log for the complete log.
         "$ref": "#/definitions/EndpointStatusLog"
+      healthz:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: I'm all for using `healthz` in the API as it has become a convention but this looks ugly for internal member names.

File: api/v1/openapi.yaml [hunk 7]
```diff
@@ -835,6 +838,29 @@ definitions:
       - ready
       - disconnecting
       - disconnected
+  EndpointHealthz:
+    description: Health of the endpoint
+    type: object
+    properties:
+      health:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: `overallHealth`?

File: api/v1/openapi.yaml [hunk 8]
```diff
@@ -835,6 +838,29 @@ definitions:
...
+       "$ref": "#/definitions/EndpointHealthzStatus"
+      bpf:
+       "$ref": "#/definitions/EndpointHealthzStatus"
+      policy:
+       "$ref": "#/definitions/EndpointHealthzStatus"
+      connected:
+        description: Is this endpoint reachable
+        type: boolean
+  EndpointHealthzStatus:
+    description: A common set of statuses for endpoint health
```
Comment (tgraf) [inline api/v1/openapi.yaml]: Needs an explanation of the different states

File: api/v1/openapi.yaml [hunk 9]
```diff
@@ -835,6 +838,29 @@ definitions:
...
+    type: string
+    enum:
+      - OK
+      - Bootstrap
+      - Pending
```
Comment (tgraf) [inline api/v1/openapi.yaml:886]: Thinking more about this, I think we want to be a bit more fine grained and split this into:
- ChangesPending
- CreationPending

File: pkg/endpoint/endpoint.go (hunk 5 above)
Comment (raybejjani) [reply pkg/endpoint/endpoint.go]: You are correct. fixed.

File: pkg/endpoint/endpoint.go (hunk 1 above)
Comment (raybejjani) [reply pkg/endpoint/endpoint.go]: Yeah, this was a bad idea. I expected more variation. I collapsed everything into a single switch.

**Comment by raybejjani** [comment]
@aanm I forgot to set the connected flag to anything but true. That is now fixed and the function is much smaller.

**Comment by raybejjani** [comment]
Soo, for all the folks in bpf, ci, docs, vendor and vagrant: I had a git/github booboo and it added you all to review. I think only api and janitors must review this. cli probably should review.

**Review by aanm — APPROVED** [review]

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by raybejjani** [comment]
I put this into wip since I didn't commit the cli files and I deleted them :/ I will rebuilt them

File: api/v1/openapi.yaml (hunk 9 above)
Comment (raybejjani) [reply api/v1/openapi.yaml:886]: Isn't bootstrap CreationPending? I can change the name on that.

File: api/v1/openapi.yaml (hunk 8 above)
Comment (raybejjani) [reply api/v1/openapi.yaml]: I'm not sure that I did it correctly, there doesn't seem to be an official way to do this (https://swagger.io/docs/specification/2-0/enums/ ?)

File: api/v1/openapi.yaml (hunk 7 above)
Comment (raybejjani) [reply api/v1/openapi.yaml]: done

**Comment by raybejjani** [comment]
Ok. Added the missing command back in, and renamed things so there are less zzz. I did not add the new states tgraf suggested.

cc @joestringer in case you wanted to see how the command was added.

File: api/v1/openapi.yaml (hunk 9 above)
Comment (tgraf) [reply api/v1/openapi.yaml:886]: OK, I'm fine with that then

**Review by tgraf — APPROVED** [review]

**Review by ianvernon — APPROVED** [review]

**Pull request merged**